### PR TITLE
HW05 — tests for HW02 and HW03

### PR DESCRIPTION
**Pull request by Captain-Vorobey**

### CODE REVIEW

**Comment by vladhilko** [comment]
Could you run `rubocop-rspec` https://github.com/rubocop/rubocop-rspec ? and add a little bit more border cases? One example can't cover all your logic, only success path

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
+require('../my_array')
+
+RSpec.describe MyArray do
+  my_arr = MyArray.new([1, 2, 3, 4, 5])
+  empty_arr = []
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:5]:
```suggestion
  subject { MyArray.new(array)}
  let(:empty_arr) { [] }
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  describe '#select' do
+    it 'returns an array of trueth values' do
+      trueth_values = [2, 4]
+      expect(my_arr.select { |el| el.even? }).to eq trueth_values
+    end
+
+    it "returns NoMethodError when block wasn't given" do
+      expect { my_arr.select }.to raise_error(NoMethodError)
+    end
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:15]:
```suggestion
    context 'when block wasn't given' do
      it "returns NoMethodError" do
        expect { my_arr.select }.to raise_error(NoMethodError)
      end
    end
```

similarly for others

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+RSpec.describe MyArray do
+  my_arr = MyArray.new([1, 2, 3, 4, 5])
+  empty_arr = []
+
+  describe '#select' do
+    it 'returns an array of trueth values' do
+      trueth_values = [2, 4]
+      expect(my_arr.select { |el| el.even? }).to eq trueth_values
+    end
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:11]:
```suggestion
  describe '#select' do
    let(:array) { [1,2,3,4,5] }
    
    context '.....' do
      let(:expected_array) { [2, 4] }
      it 'returns an array of trueth values' do
        expect(my_arr.select { |el| el.even? }).to eq expected_array
      end
    end
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    it "returns NoMethodError when block wasn't given" do
+      expect { my_arr.select }.to raise_error(NoMethodError)
+    end
+
+    it 'returns an empty array when we got empty arr in input' do
+      expect(empty_arr.select { |el| el.even? }).to be_empty
+    end
+
+    it "returns the original array with all original values when we don't have a conditional" do
+      res = [1, 2, 3, 4, 5]
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:22]: same

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 5]
```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+
+require '../lib/logs_parser'
+
+RSpec.describe LogsParser do
+  # Input data of task_1 and task_2 tests
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:6]:
```suggestion
```
delete all comments please

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  logs = <<~LOGS
+    10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] 'POST /test/2/messages HTTP/1.1' 200 48 0.0498
+    10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] 'POST /test/2/run HTTP/1.1' 200 - 0.2277
+    2018-04-23 20:30:42: SSL ERROR, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System error: Undefined error: 0 - 0>
+    10.6.246.101 - - [23/Apr/2018:20:31:39 +0300] 'POST /test/2/messages HTTP/1.1' 200 48 0.0290
+  LOGS
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:12]: use let

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
+require('../my_array')
+
+RSpec.describe MyArray do
+  subject { MyArray.new([1, 2, 3, 4, 5]) }
+
+  let(:empty_arr) { [] }
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:6]:
```suggestion
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    end
+
+    context 'when block wasn\'t given' do
+      it 'returns NoMethodError' do
+        expect { subject.select }.to raise_error(NoMethodError)
+      end
+    end
+
+    context 'when passed an empty arr' do
+      it 'returns an empty array' do
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:26]:
```suggestion
    context 'when passed an empty arr' do
      let(:empty_arr) { [] }
      
      it 'returns an empty array' do
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  describe '#each' do
+    context 'when passed valid block' do
+      it 'returns an original array' do
+        res = [1, 2, 3, 4, 5]
+        expect(subject.each { |el| el *= 100 }).to eq res
+      end
+    end
+
+    context 'when passed an empty arr' do
+      it 'returns an empty array' do
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:47]:
```suggestion
    context 'when passed an empty arr' do
      let(:empty_arr) { [] }
      
      it 'returns an empty array' do
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    end
+
+    context 'when block wasn\'t given' do
+      it 'returns NoMethodError' do
+        expect { subject.map }.to raise_error(NoMethodError)
+      end
+    end
+
+    context 'when passed an empty arr' do
+      it 'returns an empty array' do
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:74]:
```suggestion
    context 'when passed an empty arr' do
      let(:empty_arr) { [] }
      
      it 'returns an empty array' do
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+      it 'returns NoMethodError' do
+        expect { subject.each }.to raise_error(NoMethodError)
+      end
+    end
+  end
+
+  describe '#map' do
+    context 'when passed valid block' do
+      it 'returns a new array' do
+        new_arr = [100, 200, 300, 400, 500]
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:62]:
```suggestion
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    context 'when block wasn\'t given' do
+      it 'returns NoMethodError' do
+        expect { subject.each }.to raise_error(NoMethodError)
+      end
+    end
+  end
+
+  describe '#map' do
+    context 'when passed valid block' do
+      it 'returns a new array' do
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:61]:
```suggestion
    context 'when passed valid block' do
      let(:expected_array) { [100, 200, 300, 400, 500] }
      
      it 'returns a new array' do
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        new_arr = [100, 200, 300, 400, 500]
+        expect(subject.map { |el| el *= 100 }).to eq new_arr
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:63]:
```suggestion
        expect(subject.map { |el| el *= 100 }).to eq expected_array
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+      end
+    end
+
+    context 'when passed an empty arr' do
+      it 'returns an empty array' do
+        expect(empty_arr.each { |el| el *= 100 }).to be_empty
+      end
+    end
+
+    context 'when block wasn\'t given' do
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:52]:
```suggestion
    context 'when block was not given' do
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+      it "returns the original array with all original values when we don't have a conditional" do
+        expect(subject.select { |el| el + 1 }).to eq(result_array)
+      end
+    end
+  end
+
+  describe '#each' do
+    context 'when passed valid block' do
+      it 'returns an original array' do
+        res = [1, 2, 3, 4, 5]
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:41]: same

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 16]
```diff
@@ -0,0 +1,79 @@
+require('../my_array')
+
+RSpec.describe MyArray do
+  subject { MyArray.new([1, 2, 3, 4, 5]) }
+
+  let(:empty_arr) { [] }
+
+  describe '#select' do
+    let(:result_array) { [1, 2, 3, 4, 5] }
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:9]:
```suggestion
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    context 'when passed an empty arr' do
+      it 'returns an empty array' do
+        expect(empty_arr.select { |el| el.even? }).to be_empty
+      end
+    end
+
+    context 'when passed invalid block' do
+      it "returns the original array with all original values when we don't have a conditional" do
+        expect(subject.select { |el| el + 1 }).to eq(result_array)
+      end
```
Comment (B0RlS) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:34]:
```suggestion
    context 'when passed invalid block' do
      let(:expected_array) { [1, 2, 3, 4, 5] }
      
      it "returns the original array with all original values when we don't have a conditional" do
        expect(subject.select { |el| el + 1 }).to eq(expected_array)
      end
```

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    ['2018-04-23 17:17:49.7',
+     '2018-04-23 17:18:38.8',
+     '2018-04-23 17:18:39.8']
+  end
+
+  let(:time_interval) { [49.1, 1.0] }
+
+  let(:first_word) { 'hello1234' }
+  let(:second_word) { 'hello' }
+  let(:third_word) { '231!' }
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:35]:
```suggestion
```

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+  let(:first_hash) { { letters: 5, digits: 4 } }
+  let(:second_hash) { { letters: 5, digits: 0 } }
+  let(:third_hash) { { letters: 0, digits: 3 } }
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:39]:
```suggestion
```

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+      it 'returns TypeError' do
+        expect do
+          LogsParser.task_3([81, true, 'str111', :symbol, 32, { name: 'Egor', age: 20 }])
+        end.to raise_error(TypeError)
+      end
+    end
+  end
+
+  describe '.task_4' do
+    context 'when string was passed' do
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:104]:
```suggestion
  describe '.task_4' do
    let(:first_word) { 'hello1234' }
    let(:second_word) { 'hello' }  
    let(:third_word) { '231!' }
    
    let(:expected_hash1) { { letters: 5, digits: 4 } }
    let(:expected_hash2) { { letters: 5, digits: 0 } }
    let(:expected_hash3) { { letters: 0, digits: 3 } }
    
    context 'when string was passed' do
```

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+      it 'returns count of letters and digits of input word' do
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:105]:
```suggestion
      it 'returns count of letters and digits of input word', :aggregate_failures do
```

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+
+  let(:error) do
+    '2018-04-23 20:30:42: SSL ERROR, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System error: Undefined error: 0 - 0>'
+  end
+
+  let(:formatted) do
+    ['23/Apr/2018:20:30:39 +0300 FROM: 10.6.246.103 TO: /TEST/2/MESSAGES',
+     '23/Apr/2018:20:30:42 +0300 FROM: 10.6.246.101 TO: /TEST/2/RUN',
+     '23/Apr/2018:20:31:39 +0300 FROM: 10.6.246.101 TO: /TEST/2/MESSAGES']
+  end
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:23]:
```suggestion
```

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+    context 'when arr wasn\'n passed' do
+      it 'returns NoMethodError' do
+        expect { LogsParser.task_1([1, true, 'str', 4, 5]) }.to raise_error(NoMethodError)
+      end
+    end
+  end
+
+  describe '.task_2' do
+    context 'when logs were passed' do
+      it 'returns an array of formatted strings containing information about post requests' do
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:63]:
```suggestion
    context 'when logs were passed' do
      let(:formatted) do
        ['23/Apr/2018:20:30:39 +0300 FROM: 10.6.246.103 TO: /TEST/2/MESSAGES',
        '23/Apr/2018:20:30:42 +0300 FROM: 10.6.246.101 TO: /TEST/2/RUN',
        '23/Apr/2018:20:31:39 +0300 FROM: 10.6.246.101 TO: /TEST/2/MESSAGES']
      end
      
      it 'returns an array of formatted strings containing information about post requests' do
```

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+      10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] 'POST /test/2/messages HTTP/1.1' 200 48 0.0498
+      10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] 'POST /test/2/run HTTP/1.1' 200 - 0.2277
+      2018-04-23 20:30:42: SSL ERROR, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System error: Undefined error: 0 - 0>
+      10.6.246.101 - - [23/Apr/2018:20:31:39 +0300] 'POST /test/2/messages HTTP/1.1' 200 48 0.0290
+    LOGS
+  end
+
+  let(:error) do
+    '2018-04-23 20:30:42: SSL ERROR, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System error: Undefined error: 0 - 0>'
+  end
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:17]:
```suggestion
```

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+  let(:second_word) { 'hello' }
+  let(:third_word) { '231!' }
+
+  let(:first_hash) { { letters: 5, digits: 4 } }
+  let(:second_hash) { { letters: 5, digits: 0 } }
+  let(:third_hash) { { letters: 0, digits: 3 } }
+
+  describe '.task_1' do
+    context 'when arr of logs was passed' do
+      it 'returns the first string with error' do
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:43]:
```suggestion
    context 'when arr of logs was passed' do
      let(:error) do
        '2018-04-23 20:30:42: SSL ERROR, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System error: Undefined error: 0 - 0>'
      end
      
      it 'returns the first string with error' do
```

File: HW03/vorobyov_egor/spec/parsing_functions_spec.rb [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+  let(:logs) do
+    ['2018-04-23 17:17:49.7 ubuntu-xenial[14319] Debug - Calling core with action: event',
+     '2018-04-23 17:17:49.7 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101',
+     '2018-04-23 17:17:49.8 ubuntu-xenial[14319] Debug - docker event processed',
+     '2018-04-23 17:18:19.5 ubuntu-xenial[14319] Debug - monitoring grid communication health',
+     '2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - Calling core with action: messages',
+     '2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101',
+     '2018-04-23 17:18:59.8 ubuntu-xenial[14319] Debug - inside docker_handle_event']
+  end
+  let(:error_string) do
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/parsing_functions_spec.rb:16]:
```suggestion
  let(:string_with_error) do
```

File: HW03/vorobyov_egor/spec/parsing_functions_spec.rb [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+    '2018-04-23 20:30:42: SSL ERROR, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System error: Undefined error: 0 - 0>'
+  end
+  let(:no_error_string) { '10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] \'POST /test/2/run HTTP/1.1\' 200 - 0.2277' }
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/parsing_functions_spec.rb:19]:
```suggestion
  let(:string_without_error) { '10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] \'POST /test/2/run HTTP/1.1\' 200 - 0.2277' }
```

File: HW03/vorobyov_egor/spec/parsing_functions_spec.rb [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+  let(:output_difference) { 49.1 }
+  let(:core_logs) do
+    ['2018-04-23 17:17:49.7',
+     '2018-04-23 17:18:38.8']
+  end
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/parsing_functions_spec.rb:24]:
```suggestion
```

File: HW03/vorobyov_egor/spec/parsing_functions_spec.rb [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+  describe '.formatted' do
+    context 'when string passed'
+    it 'returns a formatted string' do
+      expect(ParsingFunctions.formatted(string)).to eq output_string
+    end
+  end
+
+  describe '.calling_core_logs' do
+    context 'when logs passed' do
+      it 'returns logs which calling core with action' do
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/parsing_functions_spec.rb:38]:
```suggestion
    context 'when logs passed' do
      let(:core_logs) do
        ['2018-04-23 17:17:49.7',
        '2018-04-23 17:18:38.8']
      end
      
      it 'returns logs which calling core with action' do
```

File: HW03/vorobyov_egor/spec/parsing_functions_spec.rb [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+        expect(ParsingFunctions.dates_difference(first_date, second_date)).to eq output_difference
+      end
+    end
+  end
+
+  describe '.with_error?' do
+    context 'when string has an error' do
+      it 'returns true' do
+        expect(ParsingFunctions.with_error?(error_string)).to eq true
+        expect(ParsingFunctions.with_error?(no_error_string)).to eq false
```
Comment (B0RlS) [inline HW03/vorobyov_egor/spec/parsing_functions_spec.rb:56]:
```suggestion
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+    context 'when passed valid block' do
+      let(:expected_array) { [2, 4] }
+
+      it 'returns an array of trueth values' do
+        expect(subject.select { |el| el.even? }).to eq expected_array
+      end
+    end
+
+    context 'when block was not given' do
+      it 'returns NoMethodError' do
```
Comment (vladhilko) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:18]:
```suggestion
      it 'raises NoMethodError' do
```

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 32]
```diff
@@ -0,0 +1,83 @@
...
+    context 'when passed an empty arr' do
+      let(:empty_arr) { [] }
+
+      it 'returns an empty array' do
+        expect(empty_arr.each { |el| el *= 100 }).to be_empty
+      end
+    end
+
+    context 'when block was not given' do
+      it 'returns NoMethodError' do
```
Comment (vladhilko) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:54]: and the same everywhere

File: HW02/vorobyov_egor/spec/my_array_spec.rb [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+  subject { MyArray.new([1, 2, 3, 4, 5]) }
+
+  let(:result_array) { [1, 2, 3, 4, 5] }
+
+  describe '#select' do
+    context 'when passed valid block' do
+      let(:expected_array) { [2, 4] }
+
+      it 'returns an array of trueth values' do
+        expect(subject.select { |el| el.even? }).to eq expected_array
```
Comment (vladhilko) [inline HW02/vorobyov_egor/spec/my_array_spec.rb:13]: No need to use `expected` variable for such a short result 
```suggestion
      it 'returns an array of trueth values' do
        expect(subject.select { |el| el.even? }).to eq [2, 4]
```

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 34]
```diff
@@ -0,0 +1,124 @@
...
+    context 'when arr wasn\'n passed' do
+      it 'returns NoMethodError' do
+        expect { LogsParser.task_1([1, true, 'str', 4, 5]) }.to raise_error(NoMethodError)
+      end
+    end
+  end
+
+  describe '.task_2' do
+    context 'when logs were passed' do
+      let(:formatted) do
```
Comment (vladhilko) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:49]: We should use a more clear name
```suggestion
      let(:formatted_logs) do
```

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 35]
```diff
@@ -0,0 +1,124 @@
...
+      it 'returns NoMethodError' do
+        expect { LogsParser.task_2([8, false, 'str', :symbol, 5]) }.to raise_error(NoMethodError)
+      end
+    end
+  end
+
+  describe '.task_3' do
+    context 'when arr was passed' do
+      it 'returns an array of duration betwenn 1st and 2nd, 2nd and 3rd and etc' do
+        expect(LogsParser.task_3(time_arr)).to eq time_interval
```
Comment (vladhilko) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:76]: you use `time_interval` only here, let's define this variable only inside this context in this case

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 36]
```diff
@@ -0,0 +1,124 @@
...
+
+    let(:first_hash) { { letters: 5, digits: 4 } }
+    let(:second_hash) { { letters: 5, digits: 0 } }
+    let(:third_hash) { { letters: 0, digits: 3 } }
+
+    context 'when string was passed' do
+      it 'returns count of letters and digits of input word', :aggregate_failures do
+        expect(LogsParser.task_4(first_word)).to eq first_hash
+        expect(LogsParser.task_4(second_word)).to eq second_hash
+        expect(LogsParser.task_4(third_word)).to eq third_hash
```
Comment (vladhilko) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:108]: you can write short values directly inside the specs, in your case these variabels don't mean anything 
```suggestion
        expect(LogsParser.task_4('hello1234')).to eq { letters: 5, digits: 4 }
        expect(LogsParser.task_4(second_word)).to eq second_hash
        expect(LogsParser.task_4(third_word)).to eq third_hash
```

File: HW03/vorobyov_egor/spec/logs_parser_spec.rb [hunk 37]
```diff
@@ -0,0 +1,124 @@
...
+      it 'returns ArgumentError' do
+        expect { LogsParser.task_4 }.to raise_error(ArgumentError)
+      end
+    end
+
+    context 'when value wasn\'n passed' do
+      it 'returns NoMethodError' do
+        expect { LogsParser.task_4(3) }.to raise_error(NoMethodError)
+      end
+    end
```
Comment (vladhilko) [inline HW03/vorobyov_egor/spec/logs_parser_spec.rb:122]: 2 equals specs here and you pass value 
```suggestion

```

File: HW03/vorobyov_egor/spec/parsing_functions_spec.rb [hunk 38]
```diff
@@ -0,0 +1,62 @@
+require '../lib/parsing_functions'
+
+RSpec.describe ParsingFunctions do
+  describe '.formatted' do
+    let(:string) { '10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] \'POST /test/2/messages HTTP/1.1\' 200 48 0.0498' }
+
+    let(:output_string) { '23/Apr/2018:20:30:39 +0300 FROM: 10.6.246.103 TO: /TEST/2/MESSAGES' }
+
+    context 'when string passed'
+    it 'returns a formatted string' do
```
Comment (vladhilko) [inline HW03/vorobyov_egor/spec/parsing_functions_spec.rb:10]: rubocop

**Pull request merged**